frankenphp.c

```diff
@@ -663,7 +664,24 @@ static void frankenphp_register_variables(zval *track_vars_array) {
   /* In CGI mode, we consider the environment to be a part of the server
    * variables
    */
-  php_import_environment_variables(track_vars_array);
+
+  frankenphp_server_context *ctx = SG(server_context);
+
+  /* in non-worker mode we import the os environment regularly */
+  if (!ctx->has_main_request) {
+    php_import_environment_variables(track_vars_array);
+    go_register_variables(thread_index, track_vars_array);
+    return;
+  }
+
+  /* In worker mode we cache the os environment */
+  if (os_environment == NULL) {
```

Comment:
The downside of this is that it assumes the environment is immutable (it isn't, nor is it threadsafe). The environment can and does change over an application's lifetime, for example, through `putenv` or a `TZ` variable from the os. So, there needs to be some way to pick those changes up; otherwise threads will become desynchronized. There should be away to clear the cache (through restarting the workers or SIGHUP).

Reply:
Restarting a worker will refresh the environment, it's only cached for the lifetime of running a worker script.
But you're right, it's not immutable. I think the easiest solution would be to make this an optional configuration, another option would be to just cache it for a specific amount of time. We could also expose the 'restart workers' functionality through the admin api  (if that's possible) WDYT?

The main idea is to use this in production environments, since many frameworks cache any configuration there anyways. The biggest beneficiaries are servers with tons of environment variables (100+ is not unrealistic for containers)

Reply:
Well, the maximum size of environment variables (all combined, with arguments and command name) is usually ~2mb.

In other words, you may get about the same performance by simply pre-allocating an appropriately sized array instead of a zero sized array. Without profiling and just taking a guess, I suspect most of the 'speed' is lost resizing the array as adding an element to an array (without resizing) is only O(1), while resizing it will cost O(n) -- but 'many times'. Since you are basically just doing a copy here anyway -- O(n) exactly once -- and without the cache is O(n) -- exactly once -- I don't think the cache is the actual solution.

Maybe try replacing the cache with something ridiculous:

```c
zend_hash_init(&arr, 64000, NULL, ZVAL_PTR_DTOR, 0);
```

That should prevent the array from having to completely resize over-and-over again while adding elements to the hash.

Reply:
Do you mean doing this?
```
array_init_size(track_vars_array, 64000);
php_import_environment_variables(track_vars_array);
```
Because that's a lot slower (~10 times)

I'm pretty sure the slowdown comes from [php_import_environment_variables](https://github.com/php/php-src/blob/3da6818c9e1210a76fc4cc4a1f83f66ffce76293/main/php_variables.c#L629) doing this:
```
tsrm_env_lock();
{parse all environment variables to a zval}
tsrm_env_unlock();
```

But if there's a way to optimize this without caching that would be better ofc

Reply:
oh, of course there is a lock. 🤦 

But first:

> Because that's a lot slower (~10 times)

Yeah lol, that is allocating ~40mb of memory which will be slow. I doubt you need that much, but could probably get away with something much smaller. You are mostly trying to avoid the initial doubling period 1,2,4,8,16...128. It’s probably enough to set it to something like 128, which is about 64-78 keys before it will resize. Maybe even 256 if 100 environment variables are actually common—I've never counted.

> I'm pretty sure the slowdown comes from `php_import_environment_variables` doing this:

I'm reasonably confident that we cannot get around this lock. As it is necessary to prevent a `putenv` from crashing the whole application. That being said, it would benefit from an RWLock type structure as we just need to prevent writers during this critical section, concurrent reads shouldn't be an issue. So, this is probably an improvement that can be made to TSRM itself.

That lock also needs to be held over the whole iteration of the strings because getting the environment literally gets a pointer to shared memory from the OS, and not a copy (IIRC). 

All said, an appropriately sized array + better mutex should fix this speed bump better than caching. You could even reimplement it like in [litespeed](https://github.com/php/php-src/blob/c597f92be9dcd41e61ed1c470cd11cc56fb9111b/sapi/litespeed/lsapi_main.c#L214). 

PHP extensions (which is what frankenphp kinda is) can also replace built-in functions (see: `zend_hash_str_del(EG(function_table), "putenv", sizeof("putenv") - 1);`), so we could remove `putenv` and `setenv` and replace them to use `pthread_rwlock_rdlock` and `pthread_rwlock_wrlock`. This would allow us to have concurrent readers, which should clear up the bottleneck.

Reply:
I did some testing with the litespeed implementation and a ~100kb environment variable.
With litespeed and `tsrm_env_lock()`:    ~26.000 RPS
With litespeed and no lock (unsafe):     ~72.000 RPS
With litespeed and pthread_rwlock_rdlock ~72.000 RPS
With the cache:                               ~82.000 RPS

Even though the cache is still more efficient (we don't have to parse any strings), a readlock would definitely be the right choice here. This optimization should probably be done in tsrm, copying all those functions here doesn't feel quite right. 

I still think most production applications don't care about changing environment variables though. Would you be fine with something like this?
```
frankenphp {
    worker {
        file /go/src/app/testdata/worker-with-watcher.php
        cache_env
    }
}
```
On top we could also cache these `$_SERVER` vars:  `SCRIPT_FILENAME`, `SCRIPT_NAME`, `SERVER_SOFTWARE`, `GATEWAY_INTERFACE ` and the empty `AUTH_TYPE` and `REMOTE_IDENT`

Reply:
Many people use dotenv files and many implementations use `putenv` to set the environment so it becomes available with `getenv` and they manually insert them into `$_ENV`. So long as `putenv` will still update the current and other threads, it should be fine.

> This optimization should probably be done in tsrm, copying all those functions here doesn't feel quite right.

I feel like it is the right way to go until TSRM could be updated (which also needs to include Windows support, so it is a complex proposition). It also shows that the performance for other SAPIs improves and the approach is sound. 

Going straight to the cache solution would be nearly impossible to come back from. As you show, there is even more performance there, so why would people choose the more correct, but slightly slower option?

To be clear, I like the cache idea, but I worry about the above paragraph, potential issues people might open, and diagnosing those issues. Other SAPIs do not behave this way (to my knowledge), so code won't be written with this behavior in mind. In other words, if we want to move forward, we need to deeply understand the implications, things that can go wrong, as well as making sure all of the above is documented properly.

Reply:
I did some tests with `variable_order=EGPCS`, which made me realize a few things:

- `$_ENV` is currently also only imported once and locally cached on each thread. If I modify `$_ENV` in a worker thread, it will stay modified in that thread only. `$_ENV` will not change through `putenv`.
- FPM ([and Litespeed](https://github.com/php/php-src/blob/c597f92be9dcd41e61ed1c470cd11cc56fb9111b/sapi/litespeed/lsapi_main.c#L226)) actually do the same thing, they only import variables once. I tested this with the `webdevops/php-nginx:8.3` docker image. Updating the environment from the command line or via `putenv` is actually not reflected in `$_SERVER` or `$_ENV`. It's only possible to fetch variables created with `putenv()` through `getenv()` 
- FrankenPHP is actually kind of the outlier in that it refreshes environment variables everytime the `$_SERVER` variable is accessed. That's probably also why using `putenv` with FrankenPHP is more likely to lead to a crash #915 
- In worker mode `$_SERVER` is created everytime even if not accessed (at [this line](https://github.com/dunglas/frankenphp/blob/d99b16a158f6c36897c22d28f19dc31b3b7b4eda/frankenphp.c#L196))
- $_ENV being an array stored on the thread might actually also explain some weird globals behaviour in #1061
- I tested this branch with a Laravel application and it does not change .env behavior

Knowing all that, I now again think that caching environment variables is the expected default behavior (PHP already does it automatically with `$_ENV`). The `$_SERVER` variable is populated by the SAPI, so its content is more up to interpretation. Still, it would be weird that env variables in `$_SERVER` would change, but variables in `$_ENV` would not (the SAPI doesn't define the `$_ENV` behaviour).

Maybe there are apps that need to constantly update env variables, but they can still do so via `putenv` and `getenv`. Abusing `getenv` and `putenv` like a global in-memory cache also probably isn't really intended IMO (on that note: an in-memory cache across all threads would be a cool feature). 

All that said, we can still keep the option of updating env variables on every request. But I think caching them would be a more performant and safer default. WDYT?

Reply:
Yes, `$_ENV` is a normal array that is just loaded once. #1061 is an actual php bug in TSRM, but I haven't nailed it down yet. If this fixes it, I wouldn't be surprised. `putenv` shouldn't change it. However, the next request to that thread should be reflected in `$_ENV` I believe.

#915 is likely caused by some sort of race condition between go env access and php env access. We already know it uses locks, so is thread safe _in PHP_ but not from go.

Maybe an even better solution would be to replace `getenv` and `putenv` to use go's environment access so that it is entirely thread safe (since Go also uses RWLocks to prevent concurrent access).

Reply:
In other words, we should check the following behavior:

```php
<?php

$testEnv = 'TEST_ENV';

if(!isset($_ENV[$testEnv])) {
	echo '$_ENV not set' . "\n";
}

if(!getenv($testEnv)) {
	echo "env not set\n";
}

putenv("$testEnv=TEST");
```

With fpm, I get the same output every time:

```
$_ENV not set
env not set
```

With frankenphp, after the first request, I get that response only the first time, then:

```
$_ENV not set
```

So, I believe your idea of caching the environment is the correct one is correct.

Reply:
Oh you are right, I haven't thought about go. Currently any go module writing to the environment will also cause a potential crash.

The weird thing about `$_ENV` in worker mode is that it doesn't get flushed [here](https://github.com/dunglas/frankenphp/blob/9acfb8be20604d617d80207e66472c5070b3b534/frankenphp.c#L114), I'll do some more experimenting.

Reply:
And since frankenphp is all one single process, that environment variable bleeds into other workers.

Reply:
So, really, we need a merge these two approaches. Are you on discord btw? I have the same name on discord if we want to chat more realtime.

Reply:
oh, but that won't work either because if you run `putenv('HELLO=WORLD')`, you would expect `exec('echo $HELLO'))` to output `"WORLD"` ... hmm. There isn't a way to isolate the thread, I don't think. This will get messy af really quick. You'd have to intercept `exec`, `passthru`, `system` and pcntl/parallel functions. We need hooks in php-src itself to pull this off I think.

In other words:

- some changes in TSRM to allow hooking into env locks
- hooks into spawning children (in a way that isn't a BC break for extensions) to we can modify the environment of the child before it starts. This is especially important for pcntl and parallel.
- hooks into `putenv` and `getenv`. There is already the `sapi_getenv` hook, but not one for `putenv` that I can find.

So ... wow, what a can of worms.

Reply:
I wasn't at my PC yesterday, I'll maybe dig out my old discord account next time 😄